fastcrypto-tbls/src/dkg.rs

```diff
@@ -0,0 +1,404 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Some of the code below is based on code from https://github.com/celo-org/celo-threshold-bls-rs,
+// modified for our needs.
+//
+
+use crate::ecies;
+use crate::ecies::RecoveryPackage;
+use crate::polynomial::{Poly, PrivatePoly, PublicPoly};
+use crate::random_oracle::RandomOracle;
+use crate::tbls::Share;
+use crate::types::ShareIndex;
+use fastcrypto::error::FastCryptoError;
+use fastcrypto::groups::{GroupElement, HashToGroupElement};
+use fastcrypto::traits::AllowedRng;
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use std::collections::HashMap;
+
+/// Generics below use `G: GroupElement' for the group of the VSS public key, and `EG: GroupElement'
+/// for the group of the ECIES public key.
+
+// TODO: Add weights to PkiNode, and change the DKG accordingly.
+
+/// PKI node, with a unique id and its encryption public key.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct PkiNode<EG: GroupElement> {
+    pub id: ShareIndex,
+    pub pk: ecies::PublicKey<EG>,
+}
+
+pub type Nodes<EG> = Vec<PkiNode<EG>>;
+
+/// Party in the DKG protocol.
+#[derive(Clone, PartialEq, Eq)]
+pub struct Party<G: GroupElement, EG: GroupElement> {
+    id: ShareIndex,
+    nodes: Nodes<EG>,
+    ecies_sk: ecies::PrivateKey<EG>,
+    ecies_pk: ecies::PublicKey<EG>,
+    vss_sk: PrivatePoly<G>,
+    vss_pk: PublicPoly<G>,
+    threshold: u32,
+    random_oracle: RandomOracle,
+}
+
+/// [EncryptedShare] holds the ECIES encryption of a share destined to the receiver.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct EncryptedShare<EG: GroupElement> {
+    pub receiver: ShareIndex,
+    // TODO: Consider replacing with a Enc(hkdf(g^{sk_i sk_j}), share) instead of sending a random
+    // group element, or extend ECIES to work like that.
+    pub encryption: ecies::Encryption<EG>,
+}
+
+/// [DkgFirstMessage] holds all encrypted shares a dealer sends during the first phase of the
+/// protocol.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub struct FirstMessage<G: GroupElement, EG: GroupElement> {
+    pub sender: ShareIndex,
+    /// The encrypted shares created by the sender.
+    pub encrypted_shares: Vec<EncryptedShare<EG>>,
+    /// The commitment of the secret polynomial created by the sender.
+    // TODO: add a proof of possession/knowledge?
```

Comment:
good point, do we know what other production ready implementations do?

Reply:
it's protocol dependent, i'm not sure we need it; will decide how to resolve it once i have a security proof for the protocol